tracer/src/Datadog.Trace/Debugger/Expressions/ProbeProcessor.cs

```diff
@@ -0,0 +1,268 @@
+// <copyright file="ProbeProcessor.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using Datadog.Trace.Debugger.Configurations.Models;
+using Datadog.Trace.Debugger.RateLimiting;
+using Datadog.Trace.Debugger.Snapshots;
+using Datadog.Trace.Logging;
+using SnapshotProbe = Datadog.Trace.Debugger.Configurations.Models.SnapshotProbe;
+
+namespace Datadog.Trace.Debugger.Expressions
+{
+    internal class ProbeProcessor
+    {
+        protected static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor(typeof(ProbeProcessor));
+
+        private ProbeExpressionEvaluator _evaluator;
+
+        public ProbeProcessor(ProbeDefinition probe)
+        {
+            var location = probe.Where.MethodName != null
+                               ? ProbeLocation.Method
+                               : ProbeLocation.Line;
+
+            var probeType = probe switch
+            {
+                SnapshotProbe { Capture: { } } => ProbeType.Snapshot,
+                MetricProbe => ProbeType.Metric,
+                _ => ProbeType.Log
+            };
+
+            ProbeInfo = new ProbeInfo(
+                probe.Id,
+                probeType,
+                location,
+                probe.EvaluateAt,
+                Templates: probe.Segments,
+                Condition: (probe as SnapshotProbe)?.When,
+                Metric: (probe as MetricProbe)?.Value);
+        }
+
+        internal ProbeInfo ProbeInfo { get; }
+
+        private ProbeExpressionEvaluator GetOrCreateEvaluator(MethodScopeMembers scopeMembers)
+        {
+            Interlocked.CompareExchange(ref _evaluator, new ProbeExpressionEvaluator(ProbeInfo.Templates, ProbeInfo.Condition, ProbeInfo.Metric, scopeMembers), null);
+            return _evaluator;
+        }
+
+        internal bool Process<TCapture>(ref CaptureInfo<TCapture> info, DebuggerSnapshotCreator snapshotCreator)
+        {
+            ExpressionEvaluationResult evaluationResult = default;
+            try
+            {
+                switch (info.MethodState)
+                {
+                    case MethodState.BeginLine:
+                    case MethodState.BeginLineAsync:
+                    case MethodState.EntryStart:
+                    case MethodState.EntryAsync:
+                        var captureBehaviour = snapshotCreator.DefineSnapshotBehavior(info, ProbeInfo.EvaluateAt, HasCondition());
+                        if (captureBehaviour is CaptureBehaviour.Delayed or CaptureBehaviour.NoCapture)
+                        {
+                            // we need to delay the capture process
+                            return true;
```

Comment:
Isn't the early exit in the begins (Line, LineAsync, Start, StartAsync) will skip sampling the Rate Limiter for the given Probe Id (meaning, if we technically meet the rate limit on method entry, we will not know that). Is it on purpose?